README
                        WEBCIT for the Citadel System
                               version 6.71
 
   Copyright (C) 1996-2006 by the authors.  Portions written by:
	Art Cancro
	Nathan Bryant
	Wilifried Goesgens
	Nick Grossman
	Andru Luvisi
	Dave Lindquist
	Martin Mouritzen

   This program is open source software released under the terms of the GNU
   General Public License, version 2.  Please read COPYING.txt for more
   licensing information.
 
   WebCit bundles the Prototype JavaScript Framework, writen by Sam
   Stephenson [http://prototype.conio.net].  These components are licensed to
   you under the terms of an MIT-style license.

   WebCit bundles the script.aculo.us JavaScript library, written by
   Thomas Fuchs [http://script.aculo.us, http://mir.aculo.us].  These
   components are licensed to you under the terms of an MIT-style license.

   WebCit bundles the TinyMCE text editor, written by Moxiecode Systems AB
   (http://tinymce.moxiecode.com/tinymce/docs/credits.html).  This component
   is licensed to you under the terms of the GNU Lesser General Public
   License.

   The Citadel logo was designed by Lisa Aurigemma.

 
 INTRODUCTION
 ------------
 
 Citadel is a sophisticated groupware and BBS package which allows multiple
users to simultaneously access the system using a variety of user interfaces.
This package (WebCit) is a "middleware" package which presents an HTML/HTTP
user interface to the Citadel system.
 
 What this means in practice is that after you've installed WebCit, users can
access all functions of your system using any web browser.  Since this may be
the first Citadel experience for many new users, the screens have been designed
to be attractive and easy to navigate.
 
 
 INSTALLATION
 ------------
 
 Unline some web-based packages, WebCit contains its own standalone HTTP
engine.  As a result, you can get it running quickly without all that tedious
mucking about with Apache configuration files and directories.  WebCit is not
intended to replace your Apache server, however -- it *only* provides a front
end to Citadel.  If you do not have another web server running, you may run
WebCit on port 80; however, in the more likely situation that you have Apache
or some other web server listening on port 80, you must run WebCit on another
port.  The default is port 2000.
 
 To compile from source, enter the usual commands:
  ./configure --prefix=/usr/local/webcit   [or whatever directory you prefer]
  make
  make install
 
 Package/Ports Maintainers: to make webcit fit smart into LHFS-ified systems
 read on at the end of this file, Advanced configure options.

 Then to initialize it:
  cd /usr/local/webcit
  ./setup
 
 After running setup, you just point your web browser to whatever port you
specified, such as:
 
  http://your.host.name:2000
 
 ...and log in.
 
 The included "setup" program is basically just an installation helper that
asks a series of questions and then adds the appropriate line to inittab to
start WebCit.  For most installations, this will do just fine.  If you have
special circumstances, or if you'd prefer to configure WebCit manually, you
may skip the setup program.  Instead, open /etc/inittab and add an entry
something like this:
 
 wc:2345:respawn:/usr/local/webcit/webserver
 
 
 Several command-line options are also available.  Here's the usage for
the "webserver" program:
  
 webserver [-i ip_addr] [-p http_port] [-s] [-t tracefile]
           [-c] [-f] [remotehost [remoteport]]
 
   *or*
 
 webserver [-i ip_addr] [-p http_port] [-s] [-t tracefile]
           [-c] [-f] uds /your/citadel/directory
 
 Explained: 
  
  -> ip_addr: the IP address on which you wish your WebCit server to run.
     You can leave this out, in which case WebCit will listen on all
     available network interfaces.  Normally this will be the case, but if
     you are running multiple Citadel systems on one host, it can be useful.
 
  -> http_port: the TCP port on which you wish your WebCit server to run.  If
     you are installing WebCit on a dedicated server, you can use the
     standard port 80.  Otherwise, if port 80 is already occupied by some
     other web service (probably Apache), then you'll need to select a
     different port.  If you do not specify a port number, WebCit will attempt
     to use port 2000.
     
  -> tracefile: where you want WebCit to log to.  This can be a file, a
     virtual console, or /dev/null to suppress logging altogether.
 
  -> The "-c" option causes WebCit to output an extra cookie containing the
     identity of the WebCit server.  The cookie will look like this:
       Set-cookie: wcserver=your.host.name
     This is useful if you have a cluster of WebCit servers sitting behind a
     load balancer, and the load balancer has the ability to use cookies to
     keep track of which server to send HTTP requests to.
 
  -> The "-s" option causes WebCit to present an HTTPS (SSL-encrypted) web
     service.  If you want to do both HTTP and HTTPS, you can simply run two
     instances of WebCit on two different ports.

  -> The "-f" option tells WebCit that it is allowed to follow the
     "X-Forwarded-For:" HTTP headers which may be added if your WebCit service
     is sitting behind a front end proxy.  This will allow users in your "Who
     is online?" list to appear as connecting from their actual host address
     instead of the address of the proxy.  In addition, the
     "X-Forwarded-Host:" header from the front end proxy will also be honored,
     which will help to make automatically generated absolute URL's (for
     things like GroupDAV and mailing list subscriptions) correct.
 
  -> remotehost: the name or IP address of the host on which your Citadel
     server is running.  The default is "localhost".
 
  -> remoteport: the port number on which your Citadel server is running.
     The default is port 504, the IANA-designated standard port for Citadel.
 
  -> "uds" is a keyword which tells WebCit that you wish to connect to a
     Citadel server running on the same computer, rather than using a TCP/IP
     socket.  /your/citadel/directory should be set to the actual name of the
     directory in which you have Citadel installed
     (such as /usr/local/citadel).  If you run Citadel and WebCit on the same
     computer, this is recommended, as it will run much faster.
 
 
 GRAPHICS
 --------
 
 WebCit contains graphics, templates, JavaScript code, etc. which are kept
in its "static" subdirectory.  All site-specific graphics, however, are
fetched from the Citadel server.
 
 The "images" directory on a Citadel system contains these graphics.  The
ones which you may be interested in are:
 
 -> background.gif: a background texture displayed under all web pages
 -> hello.gif: your system's logo.  It is displayed along with the logon
    banner, and on the top left corner of each page.
 
 If you would like to deploy a "favicon.ico" graphic, please put it in
the static/ directory.  WebCit will properly serve it from there.
 
 
 CALENDAR SERVICE
 ----------------
 
 WebCit contains support for calendaring and scheduling.  In order to use it
you must have libical v0.24 (or newer) on your system.  You must also be
running a Citadel server with calendaring support.  The calendar service will
be automatically configured and installed if your host system supports it.
 
 WebCit also provides Kolab-compatible free/busy data for calendar clients.
Unlike the Kolab server, however, there is no need for each user to "publish"
free/busy data -- it is generated on-the-fly from the server-side calendar
of the user being queried.  Note: in order to support Kolab clients, you must
have WebCit running in HTTPS mode on port 443, because that is what Kolab
clients will be expecting.
  
 
 HTTPS (encryption) SUPPORT
 --------------------------
 
 WebCit now supports HTTPS for encrypted connections.  When a secure server
port is specified via the "-s" flag, an HTTPS service is enabled.
 
 The service will look in the "keys" directory for the following files:
 
 citadel.key   (your server's private key)
 citadel.csr   (a certificate signing request)
 citadel.cer   (your server's public certificate)
 
 If any of these files are not found, WebCit will first attempt to link to the
SSL files in the Citadel service's directory (if Citadel is running on the
same host as WebCit), and if that does not succeed, it will automatically
generate a key and certificate.
 
 It is up to you to decide whether to use an automatically generated,
self-signed certificate, or purchase a certificate signed by a well known
authority.


 INTEGRATING INTO APACHE
 -----------------------

 If you want to include webcit into an Apache, because of you want to have it
running alongside your i.e. php applications without adding another ip or using a 
different port than 443 / 80 (which may be blocked by company firewalls for 
example) you can do this with apaches mod_proxy. One can do this either to make
webcit fill up into a full vhost (like webcit.mydomain.com) or to go into a 
subdirectory (like www.mydomain.com/webcit). find two config snippets, that can
be added to apaches sample ssl config (which you may find in 
/usr/share/doc/apache2/examples/ssl.conf for example) Warning: you should be somewhat
experienced with configuring apache to do that!

VHOST context:
<VirtualHost mydomain.com:443>
	#here some of your config stuff like logging, serveradmin...
	NameVirtualHost www.mydomain.com
    <location />
         allow from all
    </location>

    ProxyPass / http://127.0.0.1:2000/
    ProxyPassReverse / http://127.0.0.1:2000/
# make apache give away static content like images...
    alias /static /var/lib/citadel/www/static

</VirtualHost>

SUBDIR context:
<VirtualHost mydomain.com:443>
	#here some of your config stuff like logging, serveradmin...
	NameVirtualHost www.mydomain.com
    <location /webcit>
      allow from all
    </location>
    <location /listsub>
      allow from all
    </location>
    <location /groupdav>
      allow from all
    </location>
    <location /who_inner_html>
      allow from all
    </location>

    ProxyPass /webcit/ http://127.0.0.1:2000/webcit/
    ProxyPassReverse /webcit/ http://127.0.0.1:2000/webcit/
    ProxyPass /listsub/ http://127.0.0.1:2000/listsub/
    ProxyPassReverse /listsub/ http://127.0.0.1:2000/listsub/
    ProxyPass /groupdav/ http://127.0.0.1:2000/groupdav/
    ProxyPassReverse /groupdav/ http://127.0.0.1:2000/groupdav/
    ProxyPass /who_inner_html http://127.0.0.1:2000/who_inner_html
    ProxyPassReverse /who_inner_html http://127.0.0.1:2000/who_inner_html
# make apache give away static content like images...
    alias /static /var/lib/citadel/www/static
</VirtualHost>
 
 CONCLUSION
 ----------
 
 That's all you need to know to get started.  If you have any questions or
comments, please visit UNCENSORED! BBS, the home of Citadel, at
uncensored.citadel.org.


 ADVANCED CONFIGURE OPTIONS
 --------------------------
 
 To make webcit integrate smothly into your system there are several options available.
--with-staticdir defines where webcit should put'n search its templates and images. If you
want to go with a different Installation location then the point it is accessed at runtime, 
you can use --with-staticrundir. This option is meant to ease your needs if you're going
to install the static files as 'examples' in a location like /usr/share/doc/webcit/examples, 
and enable the user to copy them over to another dir (like /var/lib/citadel/www), where 
they're accessed at runtime. (The debian instatll scripts provided with this package 
do this to preserve user changes to the template system, see debian/citadel-webcit.postinstall)

Further there are possibilities to load the tiny_mce editor into a system uniq location. Webcit 
uses this standard component compose its messages for mails and postings. Several webcit installations
that may differ in design but use the same tiny_mce (which is the default that webcit ships with)
(set --with-editordir for that, it defaults to the dir the templates go)


ChangeLog
------------------------------------------------------------------------
r4488 | ajc | 2006-04-20 23:20:10 -0400 (Thu, 20 Apr 2006) | 2 lines

Added the 'Nifty Corners' library and implemented our
tabbed dialog with it.
------------------------------------------------------------------------
r4487 | dothebart | 2006-04-19 19:05:41 -0400 (Wed, 19 Apr 2006) | 2 lines

* [debian] depend on the locale packages.

------------------------------------------------------------------------
r4486 | dothebart | 2006-04-19 17:06:03 -0400 (Wed, 19 Apr 2006) | 2 lines

* [debian] just sugest the server, as we don't need it to be installed localy.

------------------------------------------------------------------------
r4485 | dothebart | 2006-04-19 15:48:55 -0400 (Wed, 19 Apr 2006) | 2 lines

* fresh copy of the automake environment.

------------------------------------------------------------------------
r4484 | dothebart | 2006-04-19 15:47:34 -0400 (Wed, 19 Apr 2006) | 3 lines

* sort the libs in other direction
* some additions in the debian build.

------------------------------------------------------------------------
r4482 | ajc | 2006-04-19 00:01:09 -0400 (Wed, 19 Apr 2006) | 1 line

Convert es.po to UTF-8
------------------------------------------------------------------------
r4480 | ajc | 2006-04-18 23:46:38 -0400 (Tue, 18 Apr 2006) | 1 line

Added es_ES.po
------------------------------------------------------------------------
r4479 | ajc | 2006-04-18 13:01:07 -0400 (Tue, 18 Apr 2006) | 2 lines

Removed the old NLS output from the configure.in summary,
and replaced with the new one.
------------------------------------------------------------------------
r4476 | dothebart | 2006-04-16 11:36:11 -0400 (Sun, 16 Apr 2006) | 2 lines

these ones where missing.

------------------------------------------------------------------------
r4475 | dothebart | 2006-04-16 11:28:29 -0400 (Sun, 16 Apr 2006) | 2 lines

* don't call autogen in buildpackages script.

------------------------------------------------------------------------
r4474 | dothebart | 2006-04-16 11:21:12 -0400 (Sun, 16 Apr 2006) | 2 lines

* added various files generated by autogen.sh, because of problems with autogen on several systems.

------------------------------------------------------------------------
r4472 | dothebart | 2006-04-15 08:28:12 -0400 (Sat, 15 Apr 2006) | 3 lines

* though POTFILES is automatically generated on newer 
  systems from POTFILES.in, we add it for the older ones.

------------------------------------------------------------------------
r4471 | ajc | 2006-04-13 23:20:26 -0400 (Thu, 13 Apr 2006) | 1 line

Version 6.81
------------------------------------------------------------------------
r4469 | ajc | 2006-04-13 22:57:16 -0400 (Thu, 13 Apr 2006) | 1 line

webserver.c: bind to LOCALEDIR for message catalog.
------------------------------------------------------------------------
r4465 | dothebart | 2006-04-09 17:30:25 -0400 (Sun, 09 Apr 2006) | 2 lines

* fix freebsd build. 

------------------------------------------------------------------------
r4463 | ajc | 2006-04-07 13:19:13 -0400 (Fri, 07 Apr 2006) | 1 line

Updated version number to 6.80
------------------------------------------------------------------------
r4457 | dothebart | 2006-04-06 19:43:18 -0400 (Thu, 06 Apr 2006) | 2 lines

we need to call another shellscript now.

------------------------------------------------------------------------
r4456 | dothebart | 2006-04-06 18:05:35 -0400 (Thu, 06 Apr 2006) | 2 lines

more modern version.

------------------------------------------------------------------------
r4455 | ajc | 2006-04-06 17:18:24 -0400 (Thu, 06 Apr 2006) | 2 lines

Fixed the 'add new note' button not woring
when there were already no notes on the page.
------------------------------------------------------------------------
r4451 | dothebart | 2006-04-05 18:19:35 -0400 (Wed, 05 Apr 2006) | 2 lines

* missing flag for libiconv added to the linker flags.

------------------------------------------------------------------------
r4445 | ajc | 2006-04-04 11:40:47 -0400 (Tue, 04 Apr 2006) | 1 line

After sending an initial page, return to the wholist
------------------------------------------------------------------------
r4439 | ajc | 2006-03-29 23:02:33 -0500 (Wed, 29 Mar 2006) | 2 lines

Zero out the 'Display name' field when displaying the
screen to create or edit a vCard.
------------------------------------------------------------------------
r4436 | ajc | 2006-03-27 12:49:36 -0500 (Mon, 27 Mar 2006) | 3 lines

Non-AJAX move or delete of a message now reloads the
read messages loop instead of simply displaying the result of the
operation on a grey screen.
------------------------------------------------------------------------
r4434 | ajc | 2006-03-26 23:21:50 -0500 (Sun, 26 Mar 2006) | 2 lines

Remove height constraint on site logo in
iconbar, and set width constraint to 120px (like all the others).
------------------------------------------------------------------------
r4432 | ajc | 2006-03-24 23:23:08 -0500 (Fri, 24 Mar 2006) | 2 lines

Updated the README with credits for the new
bubble tooltip, and cleaned up the spelling and grammar in the new sections.
------------------------------------------------------------------------
r4431 | ajc | 2006-03-24 23:03:40 -0500 (Fri, 24 Mar 2006) | 5 lines

Calendar: selecting 'Add new event' from the banner bar
now defaults the new event's date to the first day of the
displayed month, if the displayed month is not the current
month, *OR* the exact day being displayed, if the day view
is currently being displayed.  Resolves bug #165.
------------------------------------------------------------------------
r4430 | ajc | 2006-03-24 18:01:15 -0500 (Fri, 24 Mar 2006) | 3 lines

Worked around a DOM bug in Internet Explorer that caused it
to crash if we manipulate the innerHTML of a DOM node while the page is still being
rendered.

setup.c
/*
 * $Id$
 *
 * WebCit setup utility
 * 
 * (This is basically just an install wizard.  It's not required.)
 *
 */


#include "webcit.h"
#include "webserver.h"


#ifdef HAVE_NEWT
#include <newt.h>
#endif


#define UI_TEXT		0	/* Default setup type -- text only */
#define UI_DIALOG	2	/* Use the 'dialog' program */
#define UI_SILENT	3	/* Silent running, for use in scripts */
#define UI_NEWT		4	/* Use the "newt" window library */

int setup_type;
char setup_directory[SIZ];
char init_entry[SIZ];
int using_web_installer = 0;
char suggested_url[SIZ];

/*
 * Set an entry in inittab to the desired state
 */
void set_init_entry(char *which_entry, char *new_state) {
	char *inittab = NULL;
	FILE *fp;
	char buf[SIZ];
	char entry[SIZ];
	char levels[SIZ];
	char state[SIZ];
	char prog[SIZ];

	inittab = strdup("");
	if (inittab == NULL) return;

	fp = fopen("/etc/inittab", "r");
	if (fp == NULL) return;

	while(fgets(buf, sizeof buf, fp) != NULL) {

		if (num_tokens(buf, ':') == 4) {
			extract_token(entry, buf, 0, ':', sizeof entry);
			extract_token(levels, buf, 1, ':', sizeof levels);
			extract_token(state, buf, 2, ':', sizeof state);
			extract_token(prog, buf, 3, ':', sizeof prog); /* includes 0x0a LF */

			if (!strcmp(entry, which_entry)) {
				strcpy(state, new_state);
				sprintf(buf, "%s:%s:%s:%s",
					entry, levels, state, prog);
			}
		}

		inittab = realloc(inittab, strlen(inittab) + strlen(buf) + 2);
		if (inittab == NULL) {
			fclose(fp);
			return;
		}
		
		strcat(inittab, buf);
	}
	fclose(fp);
	fp = fopen("/etc/inittab", "w");
	if (fp != NULL) {
		fwrite(inittab, strlen(inittab), 1, fp);
		fclose(fp);
		kill(1, SIGHUP);	/* Tell init to re-read /etc/inittab */
	}
	free(inittab);
}




/* 
 * Shut down the Citadel service if necessary, during setup.
 */
void shutdown_service(void) {
	FILE *infp;
	char buf[SIZ];
	char looking_for[SIZ];
	int have_entry = 0;
	char entry[SIZ];
	char prog[SIZ];

	strcpy(init_entry, "");

	/* Determine the fully qualified path name of webserver */
	snprintf(looking_for, sizeof looking_for, "%s/webserver ", setup_directory);

	/* Pound through /etc/inittab line by line.  Set have_entry to 1 if
	 * an entry is found which we believe starts webserver.
	 */
	infp = fopen("/etc/inittab", "r");
	if (infp == NULL) {
		return;
	} else {
		while (fgets(buf, sizeof buf, infp) != NULL) {
			buf[strlen(buf) - 1] = 0;
			extract_token(entry, buf, 0, ':', sizeof entry);
			extract_token(prog, buf, 3, ':', sizeof prog);
			if (!strncasecmp(prog, looking_for,
			   strlen(looking_for))) {
				++have_entry;
				strcpy(init_entry, entry);
			}
		}
		fclose(infp);
	}

	/* Bail out if there's nothing to do. */
	if (!have_entry) return;

	set_init_entry(init_entry, "off");
}


/*
 * Start the Citadel service.
 */
void start_the_service(void) {
	if (strlen(init_entry) > 0) {
		set_init_entry(init_entry, "respawn");
	}
}



void cleanup(int exitcode)
{
#ifdef HAVE_NEWT
	newtCls();
	newtRefresh();
	newtFinished();
#endif
	exit(exitcode);
}



void title(char *text)
{
	if (setup_type == UI_TEXT) {
		printf("\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n<%s>\n", text);
	}
}



int yesno(char *question)
{
#ifdef HAVE_NEWT
	newtComponent form = NULL;
	newtComponent yesbutton = NULL;
	newtComponent nobutton = NULL;
#endif
	int i = 0;
	int answer = 0;
	char buf[SIZ];

	switch (setup_type) {

	case UI_TEXT:
		do {
			printf("%s\nYes/No --> ", question);
			fgets(buf, sizeof buf, stdin);
			answer = tolower(buf[0]);
			if (answer == 'y')
				answer = 1;
			else if (answer == 'n')
				answer = 0;
		} while ((answer < 0) || (answer > 1));
		break;

	case UI_DIALOG:
		sprintf(buf, "exec %s --yesno '%s' 10 72",
			getenv("CTDL_DIALOG"),
			question);
		i = system(buf);
		if (i == 0) {
			answer = 1;
		}
		else {
			answer = 0;
		}
		break;

#ifdef HAVE_NEWT
	case UI_NEWT:
		newtCenteredWindow(76, 10, "Question");
		form = newtForm(NULL, NULL, 0);
		for (i=0; i<num_tokens(question, '\n'); ++i) {
			extract_token(buf, question, i, '\n', sizeof buf);
			newtFormAddComponent(form, newtLabel(1, 1+i, buf));
		}
		yesbutton = newtButton(10, 5, "Yes");
		nobutton = newtButton(60, 5, "No");
		newtFormAddComponent(form, yesbutton);
		newtFormAddComponent(form, nobutton);
		if (newtRunForm(form) == yesbutton) {
			answer = 1;
		}
		else {
			answer = 0;
		}
		newtPopWindow();
		newtFormDestroy(form);	

		break;
#endif

	}
	return (answer);
}

void set_value(char *prompt, char str[])
{
#ifdef HAVE_NEWT
	newtComponent form;
	char *result;
	int i;
#endif
	char buf[SIZ];
	char dialog_result[PATH_MAX];
	char setupmsg[SIZ];
	FILE *fp;

	strcpy(setupmsg, "");

	switch (setup_type) {
	case UI_TEXT:
		title("WebCit setup");
		printf("\n%s\n", prompt);
		printf("This is currently set to:\n%s\n", str);
		printf("Enter new value or press return to leave unchanged:\n");
		fgets(buf, sizeof buf, stdin);
		buf[strlen(buf) - 1] = 0;
		if (strlen(buf) != 0)
			strcpy(str, buf);
		break;

	case UI_DIALOG:
		CtdlMakeTempFileName(dialog_result, sizeof dialog_result);
		sprintf(buf, "exec %s --backtitle '%s' --inputbox '%s' 19 72 '%s' 2>%s",
			getenv("CTDL_DIALOG"),
			"WebCit setup",
			prompt,
			str,
			dialog_result);
		system(buf);
		fp = fopen(dialog_result, "r");
		if (fp != NULL) {
			fgets(str, sizeof buf, fp);
			if (str[strlen(str)-1] == 10) {
				str[strlen(str)-1] = 0;
			}
			fclose(fp);
			unlink(dialog_result);
		}
		break;

#ifdef HAVE_NEWT
	case UI_NEWT:

		newtCenteredWindow(76, 10, "WebCit setup");
		form = newtForm(NULL, NULL, 0);
		for (i=0; i<num_tokens(prompt, '\n'); ++i) {
			extract_token(buf, prompt, i, '\n', sizeof buf);
			newtFormAddComponent(form, newtLabel(1, 1+i, buf));
		}
		newtFormAddComponent(form, newtEntry(1, 8, str, 74, (const char **) &result,
					NEWT_FLAG_RETURNEXIT));
		newtRunForm(form);
		strcpy(str, result);

		newtPopWindow();
		newtFormDestroy(form);	

#endif
	}
}


void important_message(char *title, char *msgtext)
{
#ifdef HAVE_NEWT
	newtComponent form = NULL;
	int i = 0;
#endif
	char buf[SIZ];

	switch (setup_type) {

	case UI_TEXT:
		printf("\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n");
		printf("       %s \n\n%s\n\n", title, msgtext);
		printf("Press return to continue...");
		fgets(buf, sizeof buf, stdin);
		break;

	case UI_DIALOG:
		sprintf(buf, "exec %s --backtitle '%s' --msgbox '%s' 19 72",
			getenv("CTDL_DIALOG"),
			title,
			msgtext);
		system(buf);
		break;

#ifdef HAVE_NEWT
	case UI_NEWT:
		newtCenteredWindow(76, 10, title);
		form = newtForm(NULL, NULL, 0);
		for (i=0; i<num_tokens(msgtext, '\n'); ++i) {
			extract_token(buf, msgtext, i, '\n', sizeof buf);
			newtFormAddComponent(form, newtLabel(1, 1+i, buf));
		}
		newtFormAddComponent(form, newtButton(35, 5, "OK"));
		newtRunForm(form);
		newtPopWindow();
		newtFormDestroy(form);	
		break;
#endif

	}
}


void display_error(char *error_message)
{
	important_message("Error", error_message);
}

void progress(char *text, long int curr, long int cmax)
{
#ifdef HAVE_NEWT

	/* These variables are static because progress() gets called
	 * multiple times during the course of whatever operation is
	 * being performed.  This makes setup non-threadsafe, but who
	 * cares?
	 */
	static newtComponent form = NULL;
	static newtComponent scale = NULL;
#endif
	static long dots_printed = 0L;
	long a = 0;
	char buf[SIZ];
	static FILE *fp = NULL;

	switch (setup_type) {

	case UI_TEXT:
		if (curr == 0) {
			printf("%s\n", text);
			printf("..........................");
			printf("..........................");
			printf("..........................\r");
			fflush(stdout);
			dots_printed = 0;
		} else if (curr == cmax) {
			printf("\r%79s\n", "");
		} else {
			a = (curr * 100) / cmax;
			a = a * 78;
			a = a / 100;
			while (dots_printed < a) {
				printf("*");
				++dots_printed;
				fflush(stdout);
			}
		}
		break;

	case UI_DIALOG:
		if (curr == 0) {
			sprintf(buf, "exec %s --gauge '%s' 7 72 0",
				getenv("CTDL_DIALOG"),
				text);
			fp = popen(buf, "w");
			if (fp != NULL) {
				fprintf(fp, "0\n");
				fflush(fp);
			}
		} 
		else if (curr == cmax) {
			if (fp != NULL) {
				fprintf(fp, "100\n");
				pclose(fp);
				fp = NULL;
			}
		}
		else {
			a = (curr * 100) / cmax;
			if (fp != NULL) {
				fprintf(fp, "%ld\n", a);
				fflush(fp);
			}
		}
		break;

#ifdef HAVE_NEWT
	case UI_NEWT:
		if (curr == 0) {
			newtCenteredWindow(76, 8, text);
			form = newtForm(NULL, NULL, 0);
			scale = newtScale(1, 3, 74, cmax);
			newtFormAddComponent(form, scale);
			newtDrawForm(form);
			newtRefresh();
		}
		if ((curr > 0) && (curr <= cmax)) {
			newtScaleSet(scale, curr);
			newtRefresh();
		}
		if (curr == cmax) {
			newtFormDestroy(form);	
			newtPopWindow();
			newtRefresh();
		}
		break;
#endif

	}
}




/*
 * check_inittab_entry()  -- Make sure "webserver" is in /etc/inittab
 *
 */
void check_inittab_entry(void)
{
	FILE *infp;
	char buf[SIZ];
	char looking_for[SIZ];
	char question[SIZ];
	char entryname[5];
	char http_port[128];
#ifdef HAVE_OPENSSL
	char https_port[128];
#endif
	char hostname[128];
	char portname[128];
	struct utsname my_utsname;

	/* Determine the fully qualified path name of webserver */
	snprintf(looking_for, sizeof looking_for, "%s/webserver", setup_directory);

	/* If there's already an entry, then we have nothing left to do. */
	if (strlen(init_entry) > 0) {
		return;
	}

	/* Otherwise, prompt the user to create an entry. */
	snprintf(question, sizeof question,
		"There is no '%s' entry in /etc/inittab.\n"
		"Would you like to add one?",
		looking_for);
	if (yesno(question) == 0)
		return;

	snprintf(question, sizeof question,
		"On which port do you want WebCit to listen for HTTP "
		"requests?\n\nYou can use the standard port (80) if you are "
		"not running another\nweb server (such as Apache), otherwise "
		"select another port.");
	sprintf(http_port, "2000");
	set_value(question, http_port);
	uname(&my_utsname);
	sprintf(suggested_url, "http://%s:%s/", my_utsname.nodename, http_port);

#ifdef HAVE_OPENSSL
	snprintf(question, sizeof question,
		"On which port do you want WebCit to listen for HTTPS "
		"requests?\n\nYou can use the standard port (443) if you are "
		"not running another\nweb server (such as Apache), otherwise "
		"select another port.");
	sprintf(https_port, "443");
	set_value(question, https_port);
#endif

	/* Find out where Citadel is. */
	if ( (using_web_installer) && (getenv("CITADEL") != NULL) ) {
		strcpy(hostname, "uds");
		strcpy(portname, getenv("CITADEL"));
	}
	else {
		snprintf(question, sizeof question,
			"Is the Citadel service running on the same host as WebCit?");
		if (yesno(question)) {
			sprintf(hostname, "uds");
			sprintf(portname, "/usr/local/citadel");
			set_value("In what directory is Citadel installed?", portname);
		}
		else {
			sprintf(hostname, "127.0.0.1");
			sprintf(portname, "504");
			set_value("Enter the host name or IP address of your "
				"Citadel server.", hostname);
			set_value("Enter the port number on which Citadel is "
				"running (usually 504)", portname);
		}
	}

	/* Generate unique entry names for /etc/inittab */
	snprintf(entryname, sizeof entryname, "c0");
	do {
		++entryname[1];
		if (entryname[1] > '9') {
			entryname[1] = 0;
			++entryname[0];
			if (entryname[0] > 'z') {
				display_error(
				   "Can't generate a unique entry name");
				return;
			}
		}
		snprintf(buf, sizeof buf,
		     "grep %s: /etc/inittab >/dev/null 2>&1", entryname);
	} while (system(buf) == 0);
	

	/* Now write it out to /etc/inittab */
	infp = fopen("/etc/inittab", "a");
	if (infp == NULL) {
		display_error(strerror(errno));
	} else {
		fprintf(infp, "# Start the WebCit server...\n");
		fprintf(infp, "h%s:2345:respawn:%s -p%s %s %s\n",
			entryname, looking_for,
			http_port, hostname, portname);
#ifdef HAVE_OPENSSL
		fprintf(infp, "s%s:2345:respawn:%s -p%s -s %s %s\n",
			entryname, looking_for,
			https_port, hostname, portname);
#endif
		fclose(infp);
		strcpy(init_entry, entryname);
	}
}




/*
 * Figure out what type of user interface we're going to use
 */
int discover_ui(void)
{

	/* Use "dialog" if we have it */
	if (getenv("CTDL_DIALOG") != NULL) {
		return UI_DIALOG;
	}
		
#ifdef HAVE_NEWT
	newtInit();
	newtCls();
	newtDrawRootText(0, 0, "WebCit Setup");
	return UI_NEWT;
#endif
	return UI_TEXT;
}





int main(int argc, char *argv[])
{
	int a;
	char aaa[256];
	int info_only = 0;
	strcpy(suggested_url, "http://<your_host_name>:<port>/");

	/* set an invalid setup type */
	setup_type = (-1);

	/* Check to see if we're running the web installer */
	if (getenv("CITADEL_INSTALLER") != NULL) {
		using_web_installer = 1;
	}

	/* parse command line args */
	for (a = 0; a < argc; ++a) {
		if (!strncmp(argv[a], "-u", 2)) {
			strcpy(aaa, argv[a]);
			strcpy(aaa, &aaa[2]);
			setup_type = atoi(aaa);
		}
		if (!strcmp(argv[a], "-i")) {
			info_only = 1;
		}
		if (!strcmp(argv[a], "-q")) {
			setup_type = UI_SILENT;
		}
	}


	/* If a setup type was not specified, try to determine automatically
	 * the best one to use out of all available types.
	 */
	if (setup_type < 0) {
		setup_type = discover_ui();
	}
	if (info_only == 1) {
		important_message("WebCit Setup", "Welcome to WebCit setup");
		cleanup(0);
	}

	/* If we're on something BSDish then we don't have inittab */
	if (access("/etc/inittab", F_OK)) {
		important_message("Not running SysV style init",
				"WebCit Setup can only run on systems that use /etc/inittab.\n"
				"Please manually configure your startup scripts to run WebCit\n"
				"when the system is booted.\n");
		cleanup(0);
	}

	/* Get started in a valid setup directory. */
	strcpy(setup_directory, WEBCITDIR);
	if ( (using_web_installer) && (getenv("WEBCIT") != NULL) ) {
		strcpy(setup_directory, getenv("WEBCIT"));
	}
	else {
		set_value("In what directory is WebCit installed?",
			setup_directory);
	}
	if (chdir(setup_directory) != 0) {
		important_message("WebCit Setup",
			  "The directory you specified does not exist.");
		cleanup(errno);
	}

	/* See if we need to shut down the WebCit service. */
	for (a=0; a<=3; ++a) {
		progress("Shutting down the WebCit service...", a, 3);
		if (a == 0) shutdown_service();
		sleep(1);
	}

	/* Now begin. */
	switch (setup_type) {

	case UI_TEXT:
		printf("\n\n\n"
			"               *** WebCit setup program ***\n\n");
		break;

	}

	check_inittab_entry();	/* Check /etc/inittab */

	/* See if we can start the WebCit service. */
	if (strlen(init_entry) > 0) {
		for (a=0; a<=3; ++a) {
			progress("Starting the WebCit service...", a, 3);
			if (a == 0) start_the_service();
			sleep(1);
		}
		sprintf(aaa,
			"Setup is finished.  You may now log in.\n"
			"Point your web browser at %s\n", suggested_url);
		important_message("Setup finished", aaa);
	}
	else {
		important_message("Setup finished",
			"Setup is finished.  You may now start the server.");
	}

	cleanup(0);
	return 0;
}
